core/src/main/java/io/grpc/internal/ManagedChannelImpl.java

```diff
@@ -527,7 +535,37 @@ public boolean isTerminated() {
     return terminated;
   }
 
-  /*
+  @Override
+  public ConnectivityState getState(boolean requestConnection) {
+    // TODO: handle requestConnection. At this point, not really sure how.
+    return state;
+  }
+
+  @Override
+  public void notifyWhenStateChanged(ConnectivityState source, Runnable callback) {
+    if (source != state) {
```

Comment:
`notifyWhenStateChanged()` needs to be thread-safe. In fact, all methods on `ManagedChannel` are thread-safe, and that interface misses a `@ThreadSafe`.

The race between `notifyWhenStateChanged()` and `maybeUpdateState()` also needs to be correctly handled.